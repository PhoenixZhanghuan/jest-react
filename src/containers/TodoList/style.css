* {
    margin: 0;
    padding: 0;
}
.header {
    line-height: 60px;
    background: #333;
    font-size: 24px;
    color: #fff;
}
.header-content {
    width: 600px;
    margin: 0 auto;
    font-size: 24px;
    color: #fff;
}
.header-input {
    width: 360px;
    margin-top: 15px;
    float: right;
    line-height: 24px;
    border-radius: 5px;
    padding: 0 10px;
}
.undo-list-item {
    line-height: 32px;
    font-size: 16px;
    margin-bottom: 10px;
    background: #fff;
    border-left: 3px solid #629a9c;
    text-indent: 10px;
    border-radius: 3px;
}
.undo-list-delete {
    float: right;
    margin-top: 6px;
    margin-right: 6px;
    width: 20px;
    height: 20px;
    line-height: 20px;
    text-align: center;
    background: #E6E3EA;
    text-indent: 0;
    border-radius: 50%
}

.undo-list-input {
    position: relative;
    top: -1px;
    width: 300px;
    line-height: 22px;
    font-size: 14px;
    text-indent: 10px;
    outline: none;
}