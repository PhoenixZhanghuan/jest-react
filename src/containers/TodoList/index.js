import React, { Component } from "react";
import Header from "./components/Header";
import UndoList from "./components/UndoList";
import axios from "axios";
import "./style.css";
class TodoList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            undoList: [],
        };
    }

    /*
        {
            data: [{
                status: 'div',
                value: 'dell lee'
            }],
            success: true
        }
    */
    componentDidMount() {
        setTimeout(() => {
            axios
                .get("/undoList.json")
                .then((res) => {
                    this.setState({
                        undoList: res.data,
                    });
                    console.log(res.data);
                })
                .catch((e) => {
                    console.log(e);
                });
        }, 5000);
    }

    addUndoItem = (value) => {
        this.setState({
            undoList: [
                ...this.state.undoList,
                {
                    status: "div",
                    value,
                },
            ],
        });
    };

    deleteItem = (index) => {
        const newList = [...this.state.undoList];
        newList.splice(index, 1);
        this.setState({
            undoList: newList,
        });
    };

    changeStatus = (index) => {
        const newList = this.state.undoList.map((item, listIndex) => {
            if (index === listIndex) {
                return {
                    ...item,
                    status: "input",
                };
            }
            return {
                ...item,
                status: "div",
            };
        });
        this.setState({
            undoList: newList,
        });
    };

    handleBlur = (index) => {
        const newList = this.state.undoList.map((item, listIndex) => {
            if (index === listIndex) {
                return {
                    ...item,
                    status: "div",
                };
            }
            return item;
        });
        this.setState({
            undoList: newList,
        });
    };

    valueChange = (index, value) => {
        const newList = this.state.undoList.map((item, listIndex) => {
            if (index === listIndex) {
                return {
                    ...item,
                    value,
                };
            }
            return item;
        });
        this.setState({
            undoList: newList,
        });
    };

    render() {
        return (
            <div>
                <Header addUndoItem={this.addUndoItem} />
                <UndoList
                    list={this.state.undoList}
                    deleteItem={this.deleteItem}
                    changeStatus={this.changeStatus}
                    handleBlur={this.handleBlur}
                    valueChange={this.valueChange}
                />
            </div>
        );
    }
}

export default TodoList;
