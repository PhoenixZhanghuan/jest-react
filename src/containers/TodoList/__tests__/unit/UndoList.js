import React from "react";
import { shallow } from "enzyme";
import UndoList from "../../components/UndoList";
import { findTestWrapper } from "../../../../utils/testUtils";
import TodoList from "../..";

it('Header 渲染样式正常', () => {
  const wrapper = shallow(<UndoList list={[]}/>);
  expect(wrapper).toMatchSnapshot();
});

it("未完成列表当数据为空数组时 count数目为0, 列表无内容", () => {  
    const wrapper = shallow(<UndoList list={[]} />);
    const countElem = findTestWrapper(wrapper, "count");
    const listItems = findTestWrapper(wrapper, "list-item");
    expect(countElem.text()).toEqual("0");
    expect(listItems.length).toEqual(0);
});

it("未完成列表当数据有内容时 count数目显示数据长度, 列表不为空", () => {
    // const listData = ['学习Jest', '学习TDD', '学习单元测试'];
    const listData = [{
        status: 'div',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]
    const wrapper = shallow(<UndoList list={listData} />);
    const countElem = findTestWrapper(wrapper, "count");
    const listItems = findTestWrapper(wrapper, "list-item");
    expect(countElem.text()).toEqual("3");
    expect(listItems.length).toEqual(3);
});

it("未完成列表当数据有内容时 要存在删除按钮", () => {
    const listData = [{
        status: 'div',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]
    const wrapper = shallow(<UndoList list={listData} />);
    const deleteItems = findTestWrapper(wrapper, "delete-item");
    expect(deleteItems.length).toEqual(3);
});

it("未完成列表当数据有内容时 点击某个删除按钮, 会调用删除方法", () => {
    const listData = [{
        status: 'div',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<UndoList list={listData} deleteItem={fn}/>);
    const deleteItems = findTestWrapper(wrapper, "delete-item");
    deleteItems.at(index).simulate("click");
    expect(fn).toHaveBeenLastCalledWith(index)
});

it("当某一项被点击时, 触发执行changeStatus函数", () => {
    const listData = [{
        status: 'div',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<UndoList list={listData} changeStatus={fn}/>);
    const deleteItems = findTestWrapper(wrapper, "list-item");
    deleteItems.at(index).simulate("click");
    expect(fn).toHaveBeenLastCalledWith(index)
});


it("当某一项状态是status是input时，展示输入框", () => {
    const listData = [{
        status: 'input',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]

    const wrapper = shallow(<UndoList list={listData} />);
    const inputItems = findTestWrapper(wrapper, "input");
    expect(inputItems.length).toBe(1);
});

it("当某一个输入框失去焦点时, 触发执行handleBlur函数", () => {
    const listData = [{
        status: 'input',
        value: '学习Jest'
    },{
        status: 'div',
        value: '学习TDD'
    },{
        status: 'div',
        value: '学习单元测试'
    }]
    const fn = jest.fn();
    const wrapper = shallow(<UndoList list={listData} handleBlur={fn}/>);
    const inputElem = findTestWrapper(wrapper, "input");
    inputElem.simulate('blur');
    expect(fn).toHaveBeenLastCalledWith(0)
});

it("当某一个输入框变更时, 触发执行valueChange函数", () => {
    const listData = [{
        status: 'input',
        value: '学习Jest'
    }]
    const value = '学习TDD';
    const fn = jest.fn();
    const wrapper = shallow(<UndoList list={listData} valueChange={fn}/>);
    const inputElem = findTestWrapper(wrapper, "input");
    inputElem.simulate('change', {
        target: {value}
    });
    expect(fn).toHaveBeenLastCalledWith(0, value)
});


it("valueChange方法被调用时, undoList数据项value被修改", () => {
    const wrapper = shallow(<TodoList />);
    const data = [{
        status: 'input',
        value: '学习Jest'
    }]
    const value = 'dell lee'
    wrapper.setState({
        undoList: data
    });
    wrapper.instance().valueChange(0, value);
    expect(wrapper.state('undoList')[0]).toEqual({
        ...data[0],
        value
    })
});