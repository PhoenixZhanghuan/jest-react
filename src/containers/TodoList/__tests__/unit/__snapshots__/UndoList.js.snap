// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 渲染样式正常 1`] = `
<div>
  <div
    data-test="count"
  >
    0
  </div>
  <ul />
</div>
`;
